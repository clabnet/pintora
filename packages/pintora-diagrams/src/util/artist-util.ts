import { Mark, MarkAttrs, Rect, MarkTypeMap, Text, Point, Path, PathCommand, createRotateAtPoint, PointTuple } from '@pintora/core'
import { PALETTE } from './theme'

export function getBaseText(): Text['attrs'] {
  return {
    x: 0,
    y: 0,
    text: '',
    fill: PALETTE.normalDark,
  }
}

/**
 * Will point to dest
 */
export function drawArrowTo(dest: Point, baseLength: number, rad: number, attrs?: Partial<MarkAttrs>): Path {
  const { x, y } = dest
  const xOffset = (baseLength / 2) * Math.tan(Math.PI / 3)
  const p: PathCommand[] = [
    ['M', x - xOffset, y - baseLength / 2], // top
    ['L', x - xOffset, y + baseLength / 2], // bottom
    ['L', x, y], // right
    ['Z'],
  ]

  const matrix = createRotateAtPoint(x, y, rad)
  return {
    type: 'path',
    matrix,
    attrs: {
      ...(attrs || {}),
      path: p,
    },
  }
}

/**
 * Will point to dest
 */
export function drawCrossTo(dest: Point, baseLength: number, rad: number, attrs?: Partial<MarkAttrs>): Path {
  const { x, y } = dest
  const offset = baseLength / 2
  const p: PathCommand[] = [
    ['M', x - offset, y - offset],
    ['L', x + offset, y + offset],
    ['M', x + offset, y - offset],
    ['L', x - offset, y + offset],
  ]

  const matrix = createRotateAtPoint(x, y, rad)
  return {
    type: 'path',
    matrix,
    attrs: {
      ...(attrs || {}),
      path: p,
    },
  }
}

export function makeMark<T extends keyof MarkTypeMap, M extends MarkTypeMap[T]>(
  type: T,
  attrs: M['attrs'],
  other?: Partial<M>,
) {
  return {
    type,
    ...(other || {}),
    attrs,
  } as M
}

export function calcDirection(start: Point, end: Point) {
  return Math.atan((end.y - start.y) / (end.x - start.x))
}
