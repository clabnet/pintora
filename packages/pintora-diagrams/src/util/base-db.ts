import { PintoraConfig } from '@pintora/core'
import { BaseDiagramIR } from './ir'
import { ConfigParam, OverrideAction } from './style'

export class BaseDb {
  configParams: ConfigParam[] = []
  overrideConfig: Partial<PintoraConfig> = {}

  addOverrideConfig(action: OverrideAction) {
    if ('error' in action) {
      console.error(action.error)
    } else {
      this.overrideConfig = action.value
    }
  }

  getBaseDiagramIR(): BaseDiagramIR {
    return {
      configParams: this.configParams,
      overrideConfig: this.overrideConfig,
    }
  }

  clear() {
    this.configParams = []
    this.overrideConfig = {}
  }
}
