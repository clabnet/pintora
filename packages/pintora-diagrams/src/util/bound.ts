import { Mark, MarkType, MarkTypeMap, Bounds, Point } from '@pintora/core'
import { makeBounds } from './mark-positioner'

export type MarkBoundCalculator<K extends MarkType> = (mark: MarkTypeMap[K]) => BoundsWithoutSize

type BoundsWithoutSize = Omit<Bounds, 'width' | 'height'>

// TODO: should consider matrix, if there is rotate
export const MARK_BOUND_CALCULATORS: Partial<{ [K in MarkType]: MarkBoundCalculator<K> }> = {
  rect({ attrs }) {
    return {
      left: attrs.x,
      right: attrs.x + attrs.width,
      top: attrs.y,
      bottom: attrs.y + attrs.height,
    }
  },
  circle({ attrs }) {
    return {
      left: attrs.x - attrs.r,
      right: attrs.x + attrs.r,
      top: attrs.y - attrs.r,
      bottom: attrs.y + attrs.r,
    }
  },
  ellipse({ attrs }) {
    return {
      left: attrs.cx - attrs.rx,
      right: attrs.cx + attrs.rx,
      bottom: attrs.cy - attrs.ry,
      top: attrs.cy + attrs.ry,
    }
  },
  line({ attrs }) {
    return {
      left: Math.min(attrs.x1, attrs.x2),
      right: Math.max(attrs.x1, attrs.x2),
      top: Math.min(attrs.y1, attrs.y2),
      bottom: Math.max(attrs.y1, attrs.y2),
    }
  },
  path() {
    return makeBounds()
    // return calcPathBound(attrs.path as PathCommand[])
  },
}

// // TODO:
// function calcPathBound(commands: PathCommand[]): BoundsWithoutSize {
//   const bounds = makeBounds()
//   commands.forEach(command => {
//   })
//   return {
//     left: -100,
//     right: 100,
//     top: 0,
//     bottom: 100,
//   }
// }

/**
 * Calculate bounds of a list of marks
 */
export function calcBound(marks: Mark[], opts: { recursive?: boolean } = {}): Bounds {
  let left = 0
  let top = 0
  let right = 0
  let bottom = 0
  const allMarks = new Set<Mark>()
  if (opts.recursive) {
    function recursiveAddMark(m: Mark) {
      if (allMarks.has(m)) return
      allMarks.add(m)
      if ('children' in m) {
        m.children.forEach(child => recursiveAddMark(child))
      }
    }
    marks.forEach(child => recursiveAddMark(child))
  } else {
    marks.forEach(child => allMarks.add(child))
  }

  for (const mark of allMarks.values()) {
    const { type } = mark
    const calculator: MarkBoundCalculator<typeof type> = MARK_BOUND_CALCULATORS[type]
    let bound: BoundsWithoutSize = {
      left: null,
      top: null,
      right: null,
      bottom: null,
    }
    if (calculator) {
      bound = calculator(mark)
    } else {
      console.warn('[calcBound] missing calculator', type)
    }
    if (bound.left !== null) left = Math.min(bound.left, left)
    if (bound.top !== null) top = Math.min(bound.top, top)
    if (bound.right !== null) right = Math.max(bound.right, right)
    if (bound.bottom !== null) bottom = Math.max(bound.bottom, bottom)
  }
  const width = right - left
  const height = bottom - top
  return { left, top, right, bottom, width, height }
}

export function updateBoundsByPoints(bounds: Bounds, points: Point[]) {
  points.forEach(p => {
    bounds.left = Math.min(bounds.left, p.x)
    bounds.right = Math.max(bounds.right, p.x)
    bounds.top = Math.min(bounds.top, p.y)
    bounds.bottom = Math.max(bounds.bottom, p.y)
    bounds.width = bounds.right - bounds.left
    bounds.height = bounds.bottom - bounds.top
  })
  return bounds
}

export function floorValues<O>(o: O) {
  for (const k of Object.keys(o)) {
    if (o[k]) o[k] = Math.floor(o[k])
  }
  return o
}
