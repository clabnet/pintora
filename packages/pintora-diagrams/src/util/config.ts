import { ConfigParam, ConfigMeta, interpreteConfigs, PintoraConfig, safeAssign, configApi, ITheme } from '@pintora/core'

export type { ConfigParam, ConfigMeta }
export { interpreteConfigs }

export type { ParamAction, OverrideConfigAction, SetTitleAction } from '../../shared-grammars/config'

/**
 * How edges are routed
 * - 'ortho' stands for orthogonal
 */
export type EdgeType = 'polyline' | 'ortho' | 'curved'

export type DiagramConfigContext = {
  configParams: ConfigParam[]
  overrideConfig: Partial<PintoraConfig>
}

export function baseGetConfigFromGlobalConfig(
  globalConfig: PintoraConfig,
  configContext: DiagramConfigContext,
  configKey: string,
) {
  const coreConfig = safeAssign<PintoraConfig['core']>({}, globalConfig.core, configContext.overrideConfig?.core)
  return safeAssign(
    { fontFamily: coreConfig.defaultFontFamily, useMaxWidth: coreConfig.useMaxWidth },
    globalConfig[configKey] || {},
  )
}

/**
 * Configurator `getConfig` will add some extra fields
 * - `themeConfig`
 */
export type EnhancedConf<T> = T & {
  themeConfig: PintoraConfig['themeConfig']
}

export function makeConfigurator<C>(opts: {
  defaultConfig: C
  configKey: string
  getConfigFromParamDirectives?(params: ConfigParam[]): Partial<C>
  getConfigFromGlobalConfig?(
    globalConfig: PintoraConfig,
    configContext: DiagramConfigContext,
    configKey: string,
  ): Partial<C>
  getConfigFromTheme(t: ITheme, conf: C): Partial<C>
}) {
  const { configKey, defaultConfig } = opts

  function getConfig(configContext: DiagramConfigContext, extraConfig?: Partial<C>): EnhancedConf<C> {
    const globalConfig: PintoraConfig = configApi.gnernateNewConfig(configContext.overrideConfig || {})
    const conf: EnhancedConf<C> = { ...defaultConfig, themeConfig: globalConfig.themeConfig }

    safeAssign(conf, opts.getConfigFromTheme(conf.themeConfig.themeVariables, conf))

    const getConfigFromGlobalConfig = opts.getConfigFromGlobalConfig || baseGetConfigFromGlobalConfig
    safeAssign(conf, getConfigFromGlobalConfig(globalConfig, configContext, configKey))

    if (extraConfig) safeAssign(conf, extraConfig)

    safeAssign(conf, configContext.overrideConfig?.[configKey])

    if (opts.getConfigFromParamDirectives) {
      safeAssign(conf, opts.getConfigFromParamDirectives(configContext.configParams))
    }

    return conf as EnhancedConf<C>
  }

  return {
    getConfig,
  }
}

/**
 * Guess config rules based on default config field and value type
 */
export function getParamRulesFromConfig<C>(config: C) {
  const out: { [key: string]: ConfigMeta } = {}
  for (const [key, value] of Object.entries(config)) {
    const t = typeof value
    let valueType: ConfigMeta['valueType']
    if (t === 'number') {
      valueType = 'size'
    } else if (t === 'boolean') {
      valueType = 'boolean'
    } else if (t === 'string') {
      if (/color|background/.test(key.toLowerCase())) {
        valueType = 'color'
      } else {
        valueType = 'string'
      }
    }

    if (valueType) {
      out[key] = { valueType }
    }
  }
  return out
}
