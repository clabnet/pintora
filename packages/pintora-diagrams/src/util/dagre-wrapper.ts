import dagre from '@pintora/dagre'
import { getGraphBounds, LayoutEdge, LayoutGraph, LayoutNode } from './graph'

export interface IEdgeData<E = unknown> {
  onLayout(data: LayoutEdge<E>): void
}

/**
 * Some common methods dealing with dagre layout
 */
export class DagreWrapper<E extends IEdgeData = IEdgeData, N extends LayoutNode = LayoutNode> {
  constructor(public g: LayoutGraph) {}

  doLayout() {
    dagre.layout(this.g)
  }

  callNodeOnLayout() {
    const graph = this.g
    graph.nodes().forEach(function (v) {
      const nodeData = graph.node(v) as unknown as N
      if (nodeData) {
        if (nodeData.onLayout) {
          nodeData.onLayout(nodeData)
        }
      }
    })
  }

  callEdgeOnLayout() {
    const graph = this.g
    graph.edges().forEach(function (e) {
      const edgeData: LayoutEdge<E> = graph.edge(e)
      if (edgeData) {
        if (edgeData.onLayout) {
          edgeData.onLayout(edgeData)
        }
      }
    })
  }

  getGraphBounds() {
    return getGraphBounds(this.g)
  }
}
