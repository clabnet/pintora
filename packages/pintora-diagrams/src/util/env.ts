export const isDev = typeof location !== 'undefined' && location.hostname === 'localhost'
