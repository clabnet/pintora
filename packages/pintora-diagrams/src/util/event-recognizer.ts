import { IDiagramEvent, IDiagramEventRecognizer, Mark, IGraphicEvent } from '@pintora/core'

export type RecognizerRule<D = any> = {
  match(m: Mark): boolean
  createDiagramEvent(e: IGraphicEvent, m: Mark, ir: D): IDiagramEvent
}

export class BaseEventRecognizer<D> implements IDiagramEventRecognizer<D> {
  rules: RecognizerRule<D>[] = []

  recognize(e: IGraphicEvent, ir: D): undefined | IDiagramEvent<any, any> {
    let d: IDiagramEvent | undefined
    if (e.markPath) {
      for (const m of e.markPath) {
        if (m.itemId) {
          for (const rule of this.rules) {
            if (rule.match(m)) {
              d = rule.createDiagramEvent(e, m, ir)
            }
          }
          if (d) break
        }
      }
    }
    return d
  }

  addPatternRecognizerRule(pattern: RegExp, createDiagramEvent: RecognizerRule<D>['createDiagramEvent']) {
    const rule = {
      match(m: Mark) {
        return pattern.test(m.itemId)
      },
      createDiagramEvent,
    }
    this.rules.push(rule)
    return this
  }
}
