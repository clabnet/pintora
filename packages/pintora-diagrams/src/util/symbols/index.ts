import './database'
import './package'
import './node'
import './cloud' // TODO: will have to modify dager-layout to support this with content margin
