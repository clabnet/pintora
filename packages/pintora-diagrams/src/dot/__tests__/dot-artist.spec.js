import { diagramRegistry } from '@pintora/core'
import { EXAMPLES } from '@pintora/test-shared'
import { testDraw, prepareDiagramConfig, stripDrawResultForSnapshot } from '../../__tests__/test-util'
import { dotDiagram } from '../index'
import '../../util/symbols'

describe('dot-artist', () => {
  beforeAll(() => {
    prepareDiagramConfig()
    diagramRegistry.registerDiagram('dotDiagram', dotDiagram)
  })

  it('match example snapshot', () => {
    expect(stripDrawResultForSnapshot(testDraw(EXAMPLES.dot.code))).toMatchSnapshot()
  })

  it('renders different arrow type', () => {
    const code = `
    dotDiagram
    digraph Test {
      bgcolor="lightyellow";
      label="Graph Label";

      a1 -> b1 [arrowhead="box"];
      n1 -> end [arrowhead="odot"];
      a2 -> end [arrowhead="open"];
    }
    `
    expect(stripDrawResultForSnapshot(testDraw(code))).toMatchSnapshot()
  })

  it('renders different node type', () => {
    const code = `
    dotDiagram
    @param fontWeight bold
    digraph {
      ellipse [shape="ellipse"];
      circle [shape="circle"];
      diamond [shape="diamond"];
      plaintext [shape="plaintext"];
    }
    `
    expect(stripDrawResultForSnapshot(testDraw(code))).toMatchSnapshot()
  })
})
