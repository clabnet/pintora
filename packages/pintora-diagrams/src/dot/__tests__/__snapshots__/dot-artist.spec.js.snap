// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dot-artist match example snapshot 1`] = `
Object {
  "mark": Object {
    "attrs": Object {},
    "children": Array [
      Object {
        "attrs": Object {},
        "children": Array [
          Object {
            "attrs": Object {
              "stroke": "#3b4044",
            },
            "type": "rect",
          },
          Object {
            "attrs": Object {
              "fill": "#3b4044",
              "fontFamily": "Source Code Pro, sans-serif",
              "fontSize": 14,
              "fontWeight": "normal",
              "text": "Sub",
              "textAlign": "center",
              "textBaseline": "top",
            },
            "type": "text",
          },
          Object {
            "attrs": Object {
              "radius": 2,
              "stroke": "#3b4044",
            },
            "type": "rect",
          },
          Object {
            "attrs": Object {
              "fill": "purple",
              "fontFamily": "Source Code Pro, sans-serif",
              "fontSize": 14,
              "fontWeight": "normal",
              "text": "a1",
              "textAlign": "center",
              "textBaseline": "middle",
            },
            "type": "text",
          },
        ],
        "type": "group",
      },
      Object {
        "attrs": Object {},
        "children": Array [
          Object {
            "attrs": Object {
              "stroke": "#3b4044",
            },
            "type": "path",
          },
          Object {
            "attrs": Object {
              "fill": "#3b4044",
            },
            "type": "path",
          },
        ],
        "type": "group",
      },
      Object {
        "attrs": Object {},
        "children": Array [
          Object {
            "attrs": Object {
              "stroke": "blue",
            },
            "type": "path",
          },
          Object {
            "attrs": Object {
              "fill": "blue",
            },
            "type": "path",
          },
        ],
        "type": "group",
      },
      Object {
        "attrs": Object {},
        "children": Array [
          Object {
            "attrs": Object {
              "stroke": "#3b4044",
            },
            "type": "path",
          },
          Object {
            "attrs": Object {
              "fill": "#3b4044",
            },
            "type": "path",
          },
        ],
        "type": "group",
      },
      Object {
        "attrs": Object {
          "radius": 2,
          "stroke": "#3b4044",
        },
        "type": "rect",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "normal",
          "text": "b1",
          "textAlign": "center",
          "textBaseline": "middle",
        },
        "type": "text",
      },
      Object {
        "attrs": Object {
          "radius": 2,
          "stroke": "#3b4044",
        },
        "type": "rect",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "normal",
          "text": "n1",
          "textAlign": "center",
          "textBaseline": "middle",
        },
        "type": "text",
      },
      Object {
        "attrs": Object {
          "radius": 2,
          "stroke": "#3b4044",
        },
        "type": "rect",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "normal",
          "text": "end",
          "textAlign": "center",
          "textBaseline": "middle",
        },
        "type": "text",
      },
      Object {
        "attrs": Object {
          "radius": 2,
          "stroke": "#3b4044",
        },
        "type": "rect",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "normal",
          "text": "a2",
          "textAlign": "center",
          "textBaseline": "middle",
        },
        "type": "text",
      },
    ],
    "type": "group",
  },
}
`;

exports[`dot-artist renders different arrow type 1`] = `
Object {
  "mark": Object {
    "attrs": Object {},
    "children": Array [
      Object {
        "attrs": Object {
          "fill": "lightyellow",
          "stroke": "transparent",
        },
        "class": "dot__frame-bg",
        "type": "rect",
      },
      Object {
        "attrs": Object {},
        "children": Array [
          Object {
            "attrs": Object {
              "stroke": "#3b4044",
            },
            "type": "path",
          },
          Object {
            "attrs": Object {
              "fill": "#3b4044",
            },
            "type": "path",
          },
        ],
        "type": "group",
      },
      Object {
        "attrs": Object {},
        "children": Array [
          Object {
            "attrs": Object {
              "stroke": "#3b4044",
            },
            "type": "path",
          },
          Object {
            "attrs": Object {
              "fill": "lightyellow",
              "stroke": "#3b4044",
            },
            "type": "path",
          },
        ],
        "type": "group",
      },
      Object {
        "attrs": Object {},
        "children": Array [
          Object {
            "attrs": Object {
              "stroke": "#3b4044",
            },
            "type": "path",
          },
          Object {
            "attrs": Object {
              "lineCap": "round",
              "stroke": "#3b4044",
            },
            "type": "path",
          },
        ],
        "type": "group",
      },
      Object {
        "attrs": Object {
          "radius": 2,
          "stroke": "#3b4044",
        },
        "type": "rect",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "normal",
          "text": "a1",
          "textAlign": "center",
          "textBaseline": "middle",
        },
        "type": "text",
      },
      Object {
        "attrs": Object {
          "radius": 2,
          "stroke": "#3b4044",
        },
        "type": "rect",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "normal",
          "text": "b1",
          "textAlign": "center",
          "textBaseline": "middle",
        },
        "type": "text",
      },
      Object {
        "attrs": Object {
          "radius": 2,
          "stroke": "#3b4044",
        },
        "type": "rect",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "normal",
          "text": "n1",
          "textAlign": "center",
          "textBaseline": "middle",
        },
        "type": "text",
      },
      Object {
        "attrs": Object {
          "radius": 2,
          "stroke": "#3b4044",
        },
        "type": "rect",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "normal",
          "text": "end",
          "textAlign": "center",
          "textBaseline": "middle",
        },
        "type": "text",
      },
      Object {
        "attrs": Object {
          "radius": 2,
          "stroke": "#3b4044",
        },
        "type": "rect",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "normal",
          "text": "a2",
          "textAlign": "center",
          "textBaseline": "middle",
        },
        "type": "text",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "normal",
          "text": "Graph Label",
          "textAlign": "center",
          "textBaseline": "top",
        },
        "type": "text",
      },
    ],
    "type": "group",
  },
}
`;

exports[`dot-artist renders different node type 1`] = `
Object {
  "mark": Object {
    "attrs": Object {},
    "children": Array [
      Object {
        "attrs": Object {},
        "mark": Object {
          "attrs": Object {},
          "children": Array [
            Object {
              "attrs": Object {
                "cx": 31,
                "cy": 25,
                "radius": 2,
                "rx": 33,
                "ry": 15,
                "stroke": "#3b4044",
                "x": 31,
                "y": 25,
              },
              "type": "ellipse",
            },
          ],
          "type": "group",
        },
        "symbolBounds": Object {
          "bottom": 40,
          "height": 30,
          "left": 21,
          "right": 41,
          "top": 10,
          "width": 66,
        },
        "type": "symbol",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "bold",
          "text": "ellipse",
          "textAlign": "center",
          "textBaseline": "middle",
        },
        "type": "text",
      },
      Object {
        "attrs": Object {},
        "mark": Object {
          "attrs": Object {},
          "children": Array [
            Object {
              "attrs": Object {
                "r": 29,
                "radius": 2,
                "stroke": "#3b4044",
                "x": 109,
                "y": 25,
              },
              "type": "circle",
            },
          ],
          "type": "group",
        },
        "symbolBounds": Object {
          "bottom": 54,
          "height": 31,
          "left": 99,
          "right": 119,
          "top": -4,
          "width": 58,
        },
        "type": "symbol",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "bold",
          "text": "circle",
          "textAlign": "center",
          "textBaseline": "middle",
        },
        "type": "text",
      },
      Object {
        "attrs": Object {},
        "mark": Object {
          "attrs": Object {},
          "children": Array [
            Object {
              "attrs": Object {
                "path": Array [
                  Array [
                    "M",
                    195.5,
                    25,
                  ],
                  Array [
                    "m",
                    -48.68181818181818,
                    0,
                  ],
                  Array [
                    "l",
                    48.68181818181818,
                    17,
                  ],
                  Array [
                    "l",
                    48.68181818181818,
                    -17,
                  ],
                  Array [
                    "l",
                    -48.68181818181818,
                    -17,
                  ],
                  Array [
                    "Z",
                  ],
                ],
                "radius": 2,
                "stroke": "#3b4044",
              },
              "type": "path",
            },
          ],
          "type": "group",
        },
        "symbolBounds": Object {
          "bottom": 42,
          "height": 34,
          "left": 178.3181818181818,
          "right": 212.6818181818182,
          "top": 8,
          "width": 97.36363636363636,
        },
        "type": "symbol",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "bold",
          "text": "diamond",
          "textAlign": "center",
          "textBaseline": "middle",
        },
        "type": "text",
      },
      Object {
        "attrs": Object {
          "fill": "transparent",
          "radius": 2,
          "stroke": "transparent",
        },
        "type": "rect",
      },
      Object {
        "attrs": Object {
          "fill": "#3b4044",
          "fontFamily": "Source Code Pro, sans-serif",
          "fontSize": 14,
          "fontWeight": "bold",
          "text": "plaintext",
          "textAlign": "center",
          "textBaseline": "middle",
        },
        "type": "text",
      },
    ],
    "type": "group",
  },
}
`;
